Orbit extensions autoupdate

sharvilshah opened the pull request:
# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [x] Changes file added for user-visible changes in `changes/` or `orbit/changes/`.
  See [Changes files](https://fleetdm.com/docs/contributing/committing-changes#changes-files) for more information.
- [x] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [x] Manual QA for all new/changed functionality
  - For Orbit and Fleet Desktop changes:
    - [x] Manual QA must be performed in the three main OSs, macOS, Windows and Linux. (Windows testing upcoming)
    - [x] Auto-update manual QA, from released version of component to new version (see [tools/tuf/test](../tools/tuf/test/README.md)).
    
# Demo 

https://www.loom.com/share/ad5dfbd8ec204092a757737f986b772b

# Assumptions

* The extension binary name must end in `.ext`, osquery enforces this
* In yaml config, the extension name, and extension file name, must be the same: for example:

```yaml
extensions:
  hello_world:
    platform: macos
    channel: stable
```

* When pushing extension targets to TUF, they must be namespaced with `extensions/`, for instance if one is pushing an extension named `hello_world`, for platform `macos` and the `stable` channel and the `hello_world.ext` filename, the TUF targets tree would look like:
`extensions/hello_world/macos/stable/hello_world.ext`

* The extension binary (`hello_world.ext`) when `exec` with `--help` must return zero status

# How to test this branch

## Set up a local TUF

* Compile and run Fleet locally like usual
* Set up TUF using either the `main.sh` script, or combination of `create_repository.sh, run_server.sh, and gen_pkgs.sh`

```bash
~/code/fleet: SYSTEMS="macos" PKG_FLEET_URL=https://localhost:8080 PKG_TUF_URL=http://localhost:8081 ENROLL_SECRET=<secret>  GENERATE_PKG=1 ./tools/tuf/test/main.sh

# copy the updates root
~/code/fleet/test_tuf: ../build/fleetctl updates root | pbcopy
```

* Paste this roots in `fleet/orbit/pkg/update/update.go` and replace the `defaultRootKeys` and `defaultURL` constants

**Note** This is required because there seems like a bug where we default to hard-coded keys when using a local/custom TUF server (#8965 for details)
 
 * recompile/regenerate orbit package with the updated root keys
 * enroll the host in Fleet
 * push the extension targets to TUF
 * configure the `agent_options` to include the `extensions` (see example aboive)

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+		case <-r.cancel:
+			return nil
+		case <-ticker.C:
+			log.Info().Msg("calling /config API to fetch/update extensions")
```
Maybe debug level?

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+				return nil
+			}
+		}
+		ticker.Reset(r.opt.CheckInterval)
```
Why reset? Is it because the CheckInterval could change?

sharvilshah replied:
We want to consistently tick at `CheckInterval` in case extension update takes longer. Not strictly needed ehre, following a similar pattern as with flags update

zwass replied:
To confirm I understand: The intended behavior is "each update check starts 15m (or the configured interval) after completion of the last update check." As opposed to "each update check starts 15m after the start of the prior update check".

Is that right? Either way seems fine to me, just want to make sure I understand.

sharvilshah replied:
@zwass The `15m` interval (`update-interval`) is different than this `CheckInterval`.  This `CheckInterval` is configured as `60s`

This `CheckInterval` is  the interval at which orbit makes an HTTP request to `/config` endpoint to check if `extensions` have been added to `agent_options`. 

The intention here with resetting the `CheckInterval` is at we want to call `/config` API at every `CheckInterval` irrespective if the previous tick took longer for instance. A bit more discussion regarding this from a previous discussion: https://github.com/fleetdm/fleet/pull/8056#discussion_r986063310

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+	return &ExtensionRunner{
+		configFetcher: configFetcher,
+		opt:           opt,
+		cancel:        make(chan struct{}),
```
Should this be a buffered `chan` like in https://github.com/fleetdm/fleet/blob/main/orbit/pkg/update/runner.go#L72-L74? Maybe that's not necessary with the `close` approach used in `Interrupt`?

sharvilshah replied:
yep! we call `close` in `Interrupt` just like in the `FlagRunner`

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+		// Extensions from Fleet is empty
+		// this can be either because of:
+		// 1. the default state, where no extensions are configured to begin with, or
+		// 2. extensions were previously were configured, but now are deleted and reverted to empty state
```
```suggestion
		// 2. extensions were previously configured, but now are deleted and reverted to empty state
```

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+		if stat.Size() > 0 {
+			// handle case 2: create/truncate the extensions.load file and let the runner interrupt, so that
+			// osquery can't startup without the extensions that were previously loaded
+			_, err := os.Create(extensionAutoLoadFile)
```
Needs close? Maybe should set permissions to `0600`?

sharvilshah replied:
thanks! updated

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+		}
+
+		// add "extensions/" as a prefix to the target name, since that's the namespace we expect for extensions on TUF
+		target := "extensions/" + k
```
```suggestion
		target := filepath.Join("extensions", k)
```
I try to always use `filepath.Join` for cross platform support.

sharvilshah replied:
ah, we do use `filepath.Join(...)` in `path` a few lines below, this is just the target name, which the runner needs to keep track of all `Targets` it needs to potentially act on. We namespace it as `extensions/<ext_name>` here as well. Updated the comment to reflect that

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+
+		// update our view of targets
+		r.updateRunner.UpdateRunnerOptTargets(target)
+		r.updateRunner.updater.SetExtentionsTargetInfo(target, v.Platform, v.Channel, v.FileName)
```
```suggestion
		r.updateRunner.updater.SetExtensionsTargetInfo(target, v.Platform, v.Channel, v.FileName)
```

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+	}
+
+	var sb strings.Builder
+	for k, v := range data {
```
Something slightly more descriptive could be nice. Maybe this?
```suggestion
	for ext, opt := range data {
```

sharvilshah replied:
Yes, agree, updated to use slightly more descriptive names

zwass commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+
+// DoExtensionConfigUpdate calls the /config API endpoint to grab extensions from Fleet
+// It parses the extensions, computes the local hash, and writes the binary path to extension.load file
+func (r *ExtensionRunner) DoExtensionConfigUpdate() (bool, error) {
```
Can you document what the `bool` return value indicates here? I'm having a hard time understanding what it means.

zwass commented on orbit/pkg/update/update.go:
```diff
@@ -70,6 +70,15 @@ func (ts Targets) SetTargetChannel(target, channel string) {
 }
 
+// UpdateRunnerOptTargets sets/updates the TargetInfo for extension targets
+func (u *Updater) SetExtentionsTargetInfo(name, platform, channel, file string) {
```
```suggestion
func (u *Updater) SetExtensionsTargetInfo(name, platform, channel, file string) {
```

zwass commented on orbit/pkg/update/update.go:
```diff
@@ -70,6 +70,15 @@ func (ts Targets) SetTargetChannel(target, channel string) {
 	ts[target] = t
 }
 
+// UpdateRunnerOptTargets sets/updates the TargetInfo for extension targets
```
```suggestion
// SetExtensionsTargetInfo sets/updates the TargetInfo for extension targets
```

zwass commented on server/fleet/agent_options.go:
```diff
@@ -14,6 +14,8 @@ type AgentOptions struct {
 	// CommandLineStartUpFlags are the osquery CLI_FLAGS
 	CommandLineStartUpFlags json.RawMessage `json:"command_line_flags,omitempty"`
+	// Extensions are the orbit managed extensions
+	Extensions json.RawMessage `json:"extensions,omitempty"`
```
Maybe this should use the [ExtensionInfo](https://github.com/fleetdm/fleet/pull/8906/files#diff-bc31589a4fd9aed0a59f1bd77a489e2f80edc722517d3290191665bca50e33d1R228-R232) type, and then that type can be imported into the runner? If that is possible, it would allow more accurate validation.

Speaking of that, does Fleet automatically validate this with the type added here? Do we need to do anything else to support the validation?

sharvilshah replied:
okay, let me chew on this, and see if we can use `ExtensionInfo` type, better validation is always nice! I followed a similar idiom as with the flags update. 

>Speaking of that, does Fleet automatically validate this with the type added here? Do we need to do anything else to support the validation?

Yep, Fleet does do a basic validation of `extensions`

lucasmrod replied:
+1 to using `ExtensionInfo` instead of a `json.RawMessage`.

sharvilshah replied:
Yep! Will be addressing it in #9046

zwass reviewed (changes requested):
Nice! I'm glad to see how few lines of code you were able to implement it in! Some suggestions and comments are included.

What would help me gain confidence to merge would be seeing a listing of the manual test cases you tried out. I start to think about this and wonder "does it behave properly if there is no extension with the appropriate name?" "... if the extension binary won't start?" "... when the extension binary changes but the tag doesn't change?" and more.

sharvilshah commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+	type ExtensionInfo struct {
+		Platform string `json:"platform"`
+		Channel  string `json:"channel"`
+		FileName string `json:"file_name"`
```
yep, updated, and removed the `file_name`

sharvilshah commented on server/service/orbit.go:
```diff
@@ -121,50 +122,50 @@ func (svc *Service) EnrollOrbit(ctx context.Context, hardwareUUID string, enroll
 }
 
-func (svc *Service) GetOrbitFlags(ctx context.Context) (json.RawMessage, error) {
+func (svc *Service) GetOrbitConfig(ctx context.Context) (json.RawMessage, json.RawMessage, error) {
```
Okay, let me take a closer look if this can be simplified. We do return the `OrbitConfig` in the `getOrbitConfig` function

sharvilshah replied:
Yep! Will be addressing it in #9046

sharvilshah commented on server/service/orbit.go:
```diff
@@ -44,8 +44,9 @@ func (r *orbitGetConfigRequest) orbitHostNodeKey() string {
-	Flags json.RawMessage `json:"command_line_startup_flags,omitempty"`
-	Err   error           `json:"error,omitempty"`
+	Flags      json.RawMessage `json:"command_line_startup_flags,omitempty"`
+	Extensions json.RawMessage `json:"extensions,omitempty"`
```
cool, let me play around with this, and see what other changes it might require

sharvilshah replied:
Will be addressing it in #9046

sharvilshah commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,160 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+	var sb strings.Builder
+	for k, v := range data {
+		// we don't want path traversal and the like in the filename
+		if strings.Contains(v.FileName, "..") || strings.Contains(v.FileName, "/") {
```
Nice! I really like this idea about having a constant allowed/valid characters in the filename across our codebase. I think, this might bit more involved than the simple regex here. 

`fleetctl updates add` works with full absolute paths, and we will need to split the filename and handle the windows case as well. Folks who use a different/custom TUF server, also may not be using `fleetctl updates` to manage targets. 

Is this something we can tackle in a following PR which does more "hardening" and address some of our tech debt? Happy to work through this if you think we need this here.

zwass reviewed (commented):
Can you add a list of the test scenarios you've tried in the PR description? That continues to be the thing that most worries me -- I want to be able to see how thorough we've been about ensuring that this behaves as expected through a variety of scenarios.

I'd be fine with seeing some of the type changes come in a follow up PR -- if you're going to do that, can you please file an issue documenting what you intend to do?

lucasmrod commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -516,6 +518,36 @@ func main() {
 		g.Add(flagRunner.Execute, flagRunner.Interrupt)
 
+		// only setup extensions autoupdate if we have enabled updates
+		if !c.Bool("disable-updates") || c.Bool("dev-mode") {
```
Why was this added as a separate `if` clause? (instead of adding it to the existing `if` clause in line 274)

sharvilshah replied:
The one at like `274` happens much more other stuff like certs, tls proxy, enroll secret, enroll to fleet, etc is configured. And since these extensions are dynamically supplied via agent options, we need orbit to be enrolled first.

lucasmrod replied:
Gotcha. Let's add that as comment.

sharvilshah replied:
updated

lucasmrod commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -516,6 +518,36 @@ func main() {
+			_, err := updateRunner.UpdateAction()
+			if err != nil {
+				// OK, initial call may fail, ok to continue
+				log.Debug().Err(err).Msg("extensions update action failed")
```
Nit: Let's make it a `log.Info()`? (to ease troubleshooting)

sharvilshah replied:
updated

lucasmrod commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -516,6 +518,36 @@ func main() {
+			}
+
+			// call UpdateAction on the updateRunner after we have fetched extensions from Fleet
+			_, err := updateRunner.UpdateAction()
```
Am worried about what would happen if any of the non-extension targets were updated (orbit, osquery, desktop), the way it looks it would just update but continue running the old version (we'd miss the update).

Maybe this is low probability bug? (We already call `updateRunner.UpdateAction()` and restart accordingly in line 301.)

sharvilshah replied:
I wouldn't quite categorize this as a bug..? The call to `UpdateAction()` on line `301` happens at every `update-interval`, and I specifically rely on it, and since it's added to the oklog run group, it does restart things correctly. 

I am using the "extra" `UpdateAction()` call here to kick start things after orbit is enrolled and extensions are fetched from agent options.

lucasmrod replied:
Sorry, am not following:
1. Line 534 calls `UpdateAction`; say it does update Orbit, osquery or desktop. First return value is ignored.
2. Orbit continues to run as usual.
3. UpdateAction is executed every update-interval but it will not detect the new update, because it already took place.

It's an edge case for sure.

sharvilshah replied:
Ah, I see. Yeah this is an edge case. I do want to have this initial call to `UpdateAction` so that extensions can be quickly loaded. I think perhaps, I can break this function in two, and handle the case where only extension related `UpdateAction` is called, but I am thinking that this could be handled in a follow up, as a part of #9046?

lucasmrod commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -516,6 +518,36 @@ func main() {
+			extensionAutoLoadFile := filepath.Join(c.String("root-dir"), "extensions.load")
+			stat, err := os.Stat(extensionAutoLoadFile)
+			// we only want to add the extensions_autoload flag to osquery, if the file exists and size > 0
+			if !errors.Is(err, os.ErrNotExist) && stat.Size() > 0 {
```
Let's log the error if `err != nil` and `!errors.Is(err, os.ErrNotExist)`. The usual `switch` statement, something like:
```go
// We only want to add the extensions_autoload flag to osquery if the file exists and size > 0.
switch {
case err == nil:
  if stat.Size() > 0 {
    log.Debug().Msg("adding --extensions_autoload flag for file " + extensionAutoLoadFile)
    options = append(options, osquery.WithFlags([]string{"--extensions_autoload", extensionAutoLoadFile}))
  } else {
    log.Info(... empty extensions file ...)
  }
case errors.Is(err, os.ErrNotExist):
  // OK, nothing to do.
default:
  log.Error(...)
}
```

sharvilshah replied:
Ah okay, any reason to prefer `switch/case` over `if` here? Maybe it's just me, but `if` reads a bit cleaner..? Happy to change it though, since that's the pattern we use often

lucasmrod replied:
No preference, as long as we cover the mentioned case.

sharvilshah replied:
updated

lucasmrod commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -516,6 +518,36 @@ func main() {
+
+			extensionAutoLoadFile := filepath.Join(c.String("root-dir"), "extensions.load")
+			stat, err := os.Stat(extensionAutoLoadFile)
+			// we only want to add the extensions_autoload flag to osquery, if the file exists and size > 0
```
What happens if:
1. Initially there's no `extensions_autoload`, and `osqueryd` is started without it
2. Thereafter the user pushes and configures extensions
3. Will Orbit restart and run this part again? (to run `osqueryd` with the `--extensions_autoload` flag)

sharvilshah replied:
Yes, orbit will restart and run this again. I have specifically tested this case.

lucasmrod commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,180 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+		// 2. extensions were previously configured, but now are deleted and reverted to empty state
+
+		// Handle case 1, where our autoload file does not exist, so there is nothing to update and no error
+		stat, err := os.Stat(extensionAutoLoadFile)
```
Similar err check here, if there's e.g. a permissions error to load the file we won't be logging the error. Let's add a err log.

sharvilshah replied:
updated

lucasmrod commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,180 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+		filename := extensionName + ".ext"
+
+		// we don't want path traversal and the like in the filename
+		if strings.Contains(filename, "..") || strings.Contains(filename, "/") {
```
How about windows path separator? `\`

sharvilshah replied:
updated

lucasmrod commented on orbit/pkg/update/runner.go:
```diff
@@ -30,6 +30,17 @@ type Runner struct {
 }
 
+// UpdateRunnerOptTargets updates the RunnerOptions.Targets with the given target
+func (r *Runner) UpdateRunnerOptTargets(target string) {
```
Looks like there's a race now.

We should now protect `r.opt.Targets` with a lock.

sharvilshah replied:
updated, now using a mutex

lucasmrod commented on orbit/pkg/update/update.go:
```diff
@@ -70,6 +70,15 @@ func (ts Targets) SetTargetChannel(target, channel string) {
 }
 
+// SetExtensionsTargetInfo sets/updates the TargetInfo for extension targets
+func (u *Updater) SetExtensionsTargetInfo(name, platform, channel, file string) {
```
Same as `UpdateRunnerOptTargets`:

Looks like there's a race now.

We should now protect `r.opt.Targets` with a lock.

sharvilshah replied:
updated, now using a mutex

lucasmrod commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,180 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+
+		// update local hashes
+		log.Info().Msgf("updating local hash(%s)=%x", targetName, localHash)
+		r.updateRunner.localHashes[targetName] = localHash
```
This write races with the read in `r.UpdateAction`.

sharvilshah replied:
updated, now using a mutex

lucasmrod commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,180 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+
+		sb.WriteString(path + "\n")
+	}
+	if err := os.WriteFile(extensionAutoLoadFile, []byte(sb.String()), constant.DefaultFileMode); err != nil {
```
Shouldn't you first update the file, and then do the `r.updateRunner.localHashes` updates?

sharvilshah replied:
No, this is updated as part of the loop. We store the path on disk in a string buffer and write it to the extensions.load file after the loop.

lucasmrod commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,180 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+		// the path is: <root-dir>/bin/extensions/hello_world/<platform>/<channel>/hello_world.ext
+		path := filepath.Join(rootDir, "bin", "extensions", extensionName, platform, channel, filename)
+
+		meta, err := r.updateRunner.updater.Lookup(targetName)
```
You should call update metadata before Lookup? (from reading Lookup docs)

sharvilshah replied:
updated

lucasmrod commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,180 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+			return false, fmt.Errorf("unable to lookup metadata for target: %s, %w", targetName, err)
+		}
+
+		_, localHash, err := fileHashes(meta, path)
```
This will always return `os.ErrNotExists` on the first addition of the extension to the `r.updateRunner`. (Beware that the development `main.sh` script always runs with a `10s` update interval, versus the default/production value of `1h`.)

sharvilshah replied:
This is expected, and I think ok as well. We do call `UpdateAction()` after successfully fetching extensions from `/config`.  

This is more of an optimization, since `UpdateAction` will compare local and meta hashes, and we want to provide the local hash, otherwise it will redownload the extension and restart.

The longer update interval, would mean longer waiting (and I think default is 15 minutes, or probably that's just the hard-coded fallback value..?). 

Thoughts?

lucasmrod replied:
My mistake, correct, it's 15m.

--

Let's clarify that the only time `DoExtensionConfigUpdate` returns `true, nil` is when the Fleet server has cleared extensions and the host has some extensions set up.

Either document with named returns as below, or add the above sentence to the method doc.
`func (r *ExtensionRunner) DoExtensionConfigUpdate() (extensionsCleared bool, err error) {`

On the caller side this should actually be:
```go
extensionsCleared, err := r.DoExtensionConfigUpdate()
if err != nil {
  log.Info().Err(err).Msg("ext update failed")
}
if extensionsCleared {
  log.Info().Msg("extensions were cleared on the server")
  return nil
}
```
--

For the refactor issue later:
`DoExtensionConfigUpdate` is too long (thus hard to follow/review). Let's refactor it using multiple functions.

lucasmrod reviewed (commented):
First pass done. Left a couple of questions/comments.

sharvilshah commented:
Thanks @lucasmrod! Updated and addressed most of the feedback. 

@zwass: Follow-up enhancements in #9046 

Testing Strategy, this is what I have tested:

* Running this orbit with older version of Fleet Server: agent options validation will fail on `extension` as expected, preventing `extension` as a part of config.
* Bad `extension_name` or one that doesnt exist in TUF: we just log an error that target name is not found, we don't load it, and continue to run
* When the extension binary doesn't exec: we log and error (and we don't load it) and continue
* when extension binary changes and tag doesn't change: the hash comparison changes and so we update the target as usual

lucasmrod commented:
We need docs (maybe I missed them) for how admins might configure/use this feature (same for the osquery flags feature).

zwass reviewed (dismissed):
@sharvilshah can you please answer my question in the code comments?

I'm going to propose that we merge this during the freeze since the server changes are minimal (I believe we are just changing what's accepted as part of agent options). Sound good to folks?

lucasmrod commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -126,6 +126,191 @@ func (r *FlagRunner) DoFlagsUpdate() (bool, error) {
+			return false, nil
+		default:
+			// we do not want orbit to restart, just log the error
+			log.Debug().Err(err).Msg("extensionsUpdate: error stating file at " + extensionAutoLoadFile)
```
This is logging the error on `debug` mode only.

Let's return `return false, fmt.Errorf("stat file: %s", extensionAutoLoadFile)`

The caller `ExtensionRunner.Execute` will log the error.

lucasmrod commented on orbit/pkg/update/runner.go:
```diff
@@ -28,6 +29,20 @@ type Runner struct {
+			return
+		}
+	}
+	r.mu.Lock()
```
Must move to the top of the method. `r.opt.Targets`, read access must be protected too.

lucasmrod reviewed (changes requested):
Left some comments to address.

sharvilshah commented on orbit/pkg/update/flag_runner.go:
```diff
@@ -206,30 +206,34 @@ func (r *ExtensionRunner) DoExtensionConfigUpdate() (bool, error) {
+				// we want to return true here, and restart with the empty extensions.load file
+				// so that we "unload" the previously loaded extensions
+				return true, nil
+			} else {
```
This has been resolved by commit 977c7aa8a22fe6bc625a40d0b028229a7b7c1bd2, not sure why it's still showing up here, will push again

sharvilshah commented:
> We need docs (maybe I missed them) for how admins might configure/use this feature (same for the osquery flags feature).

Hey @lucasmrod, the docs for this are in a separate branch (pushing that PR shortly), and the osquery flags are documented here under "Agent Options" (https://fleetdm.com/docs/using-fleet/configuration-files#code-command-line-flags-code-option)

sharvilshah commented:
Docker seems to be timing out in CI

```
Vmnetd: Registering com.docker.vmnetd in domain Optional(kSMDomainSystemLaunchd)...
Vmnetd: Privileged installation successful
Vmnetd: isReady: ping succeeded
Privileged components installed
Waiting for Docker to start up...
Error: The action has timed out.
```

sharvilshah commented:
The test failure is because of GitHub is rate limiting us:

```
=== RUN   TestDownloadCPETranslations
    nettest.go:33: network test start: TestDownloadCPETranslations
    sync_test.go:83: 
        	Error Trace:	/home/runner/work/fleet/fleet/server/vulnerabilities/nvd/sync_test.go:83
        	Error:      	Received unexpected error:
        	            	GET https://api.github.com/repos/fleetdm/nvd/releases?per_page=10: 403 API rate limit exceeded for 52.226.82.114. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.) [rate reset in 8m50s]
        	Test:       	TestDownloadCPETranslations
    nettest.go:36: network test done: TestDownloadCPETranslations
--- FAIL: TestDownloadCPETranslations (0.02s)
FAIL
coverage: 75.7% of statements
FAIL	github.com/fleetdm/fleet/v4/server/vulnerabilities/nvd	177.668s
```

sharvilshah commented:
Re-ran the job, CI is green

zwass reviewed (approved):
We just re-discussed merging this with the minimal changes on the Fleet server side. We'll do separate QA on the agent portions before releasing.

zhumo commented:
Is there end user documentation for this?
